pkg/cloud/vsphere/services/govmomi/esxi/clone.go

```diff
@@ -17,12 +17,271 @@ limitations under the License.
 package esxi
 
 import (
+	"fmt"
+	"path"
+	"strings"
+
 	"github.com/pkg/errors"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
 
 	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/cloud/vsphere/context"
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/cloud/vsphere/services/govmomi/extra"
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/cloud/vsphere/services/govmomi/template"
 )
 
 // Clone kicks off a clone operation on ESXi to create a new virtual machine.
 func Clone(ctx *context.MachineContext, userData []byte) error {
-	return errors.New("temporarily disabled esxi support")
+	vm, err := NewVmContext(ctx)
+	if err != nil {
+		return err
+	}
+	ctx.Logger.V(6).Info("preparing esxi create spec")
+
+	diskUUIDEnabled := true
+	spec := &types.VirtualMachineConfigSpec{
+		// Use the object UID as the instanceUUID for the VM
+		InstanceUuid: string(ctx.Machine.UID),
+		Flags: &types.VirtualMachineFlagInfo{
+			DiskUuidEnabled: &diskUUIDEnabled,
+		},
+	}
+
+	mo, _ := vm.GetVirtualMachineMO()
+	spec.GuestId = mo.Config.GuestId         // set GuestId from template
+	spec.Firmware = mo.Config.Firmware       // set Firmware from template
+	spec.BootOptions = mo.Config.BootOptions // set BootOptions from Template
+
+	ctx.Logger.V(4).Info("assigned VM instance UUID from machine UID", "uid", string(ctx.Machine.UID))
+	if ctx.MachineConfig.MachineSpec.NumCPUs > 0 {
+		spec.NumCPUs = int32(vm.MachineConfig.MachineSpec.NumCPUs)
+	}
+	if ctx.MachineConfig.MachineSpec.MemoryMB > 0 {
+		spec.MemoryMB = vm.MachineConfig.MachineSpec.MemoryMB
+	}
+	spec.Annotation = vm.String()
+	spec.Name = vm.Machine.Name
+	deviceSpecs := []types.BaseVirtualDeviceConfigSpec{}
+
+	disks, err := thinCloneDisks(vm)
+	if err != nil {
+		return err
+	}
+	for _, disk := range disks {
+		deviceSpecs = append(deviceSpecs, disk)
+	}
+
+	nics, err := GetNics(vm)
+	if err != nil {
+		return err
+	}
+	for _, device := range nics {
+		deviceSpecs = append(deviceSpecs, device)
+	}
+
+	var devices object.VirtualDeviceList
+	if devices, err = GetSerialPort(vm, devices); err != nil {
+		return err
+	}
+	devicesConfigs, _ := devices.ConfigSpec(types.VirtualDeviceConfigSpecOperationAdd)
+	for _, device := range devicesConfigs {
+		deviceSpecs = append(deviceSpecs, device)
+	}
+
+	spec.DeviceChange = deviceSpecs
+	spec.Files = &types.VirtualMachineFileInfo{
+		VmPathName: fmt.Sprintf("[%s]", vm.Datastore.Name()),
+	}
+
+	var extraConfig extra.Config
+	extraConfig.SetCloudInitUserData(userData)
+	spec.ExtraConfig = extraConfig
+
+	task, err := CreateVM(vm, spec)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to create vm")
+	}
+	ctx.Logger.V(4).Info("Task submitted: " + task.Reference().Value)
+	ctx.MachineStatus.TaskRef = task.Reference().Value
+	return nil
+}
+
+// VmContext is a context that adds govmomi objects to a MachineContext
+type VmContext struct {
```

Comment:
Hi @moshloop,

Does this or any of its functions need to be exported?

Reply:
So I renamed this to GovmomiContext and moved to a new context package, a different PR will refactor vcenter to also use this